itest/rust/src/dictionary_test.rs

```diff
@@ -400,3 +372,279 @@ fn dictionary_equal() {
+    let mut dictionary2 = dictionary.share();
+    let mut iter = dictionary.iter_shared();
+
+    for _ in 0..4 {
+        for _ in 0..4 {
+            for _ in 0..16 {
+                iter.next();
+            }
+            dictionary2.insert("a", "b");
+        }
+        dictionary2.clear();
+        dictionary2.extend((0..64).zip(0..64));
+    }
+    assert_eq!(dictionary2, (0..64).zip(0..64).collect());
+}
+
+#[itest]
+fn dictionary_iter_simultaneous() {
+    let dictionary = dict! {
+        "foo": 10,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    let map: HashMap<String, (Variant, Variant)> = dictionary
+        .iter_shared()
+        .typed::<String, Variant>()
+        .zip(dictionary.iter_shared().typed::<String, Variant>())
+        .map(|((mut k1, v1), (k2, v2))| {
+            k1.push_str(k2.as_str());
+            (k1, (v1, v2))
+        })
+        .collect();
+
+    assert!(map.len() == 4);
+
+    let mut tens = 0;
+    let mut trues = 0;
+    let mut foobars = 0;
+    let mut nils = 0;
+
+    for v in map.iter().flat_map(|(_, (v1, v2))| [v1, v2]) {
+        if let Ok(b) = bool::try_from_variant(v) {
+            assert!(b);
+            trues += 1;
+        } else if let Ok(i) = i64::try_from_variant(v) {
+            assert_eq!(i, 10);
+            tens += 1;
+        } else if let Ok(s) = String::try_from_variant(v) {
+            assert_eq!(s.as_str(), "foobar");
+            foobars += 1;
+        } else {
+            assert!(v.is_nil());
+            nils += 1;
+        }
+    }
+
+    assert_eq!(tens, 2);
+    assert_eq!(trues, 2);
+    assert_eq!(foobars, 2);
+    assert_eq!(nils, 2);
+}
+
+#[itest]
+fn dictionary_iter_panics() {
+    expect_panic(
+        "Dictionary containing integer keys should not be convertible to a String iterator",
+        || {
+            let dictionary: Dictionary = (0..10).zip(0..).collect();
+            let _set: HashSet<String> = dictionary.keys_shared().typed::<String>().collect();
+        },
+    );
+
+    expect_panic(
+        "Dictionary containing integer entries should not be convertible to a (String,String) iterator",
+        || {
+            let dictionary: Dictionary = (0..10).zip(0..).collect();
+            let _set: HashMap<String,String> = dictionary.iter_shared().typed::<String,String>().collect();
+        },
+    );
+}
+
+// The below tests erase entries mid-iteration. This is not supported by godot dictionaries
+// however it shouldn't cause unsafety or panicking. Rather the outcome of the iteration is not
+// guaranteed. These tests therefore test two main things:
+// 1. The dictionary is not corrupted by erasing mid-iteration.
+// 2. Our implementation behaves the same as godot.
+//
+// #2 may change in the future, so equivalent gdscript code is provided. That way we can
+// easily check if a failure is a false negative caused by godot's behavior changing.
```

Comment:
Great description, I appreciate the documentation of thoughts that went into the tests.

Minor nitpick about capitalization: `Godot` and `GDScript` 😉
Also in some other places, maybe global search&replace is the easiest.